File: 백준/Gold/11444. 피보나치 수 6/피보나치 수 6.py
import sys 
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

MOD = 1_000_000_007

def mul(a, b):
        c = [[0]*2 for _ in range(2)]
        for i in range(2):
            for k in range(2):
                for j in range(2):
                    c[i][j] = (c[i][j]+a[i][k]*b[k][j])%MOD
        return c

def main():
    n = int(input())
    if n == 0: print(0); return
    base = [[1, 1], [1, 0]]
    ans = [[1, 0], [0, 1]]
    while n:
        if n&1: ans = mul(ans, base)
        base = mul(base, base)
        n >>= 1
    print(ans[0][1])
    return  
if __name__ == '__main__':
    main()

File: 백준/Gold/11444. 피보나치 수 6/피보나치 수 6.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;
typedef vector<vector<ll>> matrix;

const ll MOD = 1'000'000'007;

matrix mul(matrix& a, matrix& b){
    matrix ret(2, vector<ll>(2, 0));
    for(int i = 0; i < 2; i++) for(int k = 0; k < 2; k++) for(int j = 0; j < 2; j++){
        ret[i][j] = (ret[i][j]+a[i][k]*b[k][j])%MOD;
    }
    return ret;
}

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    ll n; cin >> n;
    if(n == 0) return !(cout << 0);
    matrix base = {{1, 1}, {1, 0}};
    matrix ans = {{1, 0}, {0, 1}};
    while(n){
        if(n&1) ans = mul(ans, base);
        base = mul(base, base);
        n >>= 1;
    }
    cout << ans[0][1];
    return 0;
} 